feat: improve resources_monitoring for OOM & OOD monitoring

defelmnq opened the pull request:
As requested for [this issue](https://github.com/coder/internal/issues/245) we need to have a new resource `resources_monitoring` in the agent. 

It needs to be parsed from the provisioner and inserted into a new db table.

defelmnq committed:
work on resources monitoring

defelmnq committed:
work on migrations and db

defelmnq committed:
work on database integration

defelmnq committed:
add metadata logic to db

defelmnq committed:
add metadata logic to db

defelmnq committed:
revert formatting

defelmnq committed:
improve errors handling

defelmnq committed:
work on fake db

defelmnq committed:
work on db

defelmnq committed:
update provisioner version

defelmnq committed:
work on testdata

defelmnq committed:
merge

defelmnq committed:
work on resources monitor

defelmnq committed:
worked on dbmem and dbauthz

defelmnq committed:
add dbauthz tests

defelmnq committed:
work on errors message improvement

johnstcn commented on coderd/database/migrations/000288_agent_resource_monitors.down.sql:
```diff
@@ -0,0 +1,2 @@
+DROP TABLE IF EXISTS workspace_agent_memory_resource_monitors;
+DROP TABLE IF EXISTS workspace_agent_volume_resource_monitors;
```
Obligatory reminder to double-check migration number before merging!

johnstcn commented on provisioner/terraform/testdata/calling-module/calling-module.tfplan.json:
```diff
@@ -259,7 +259,7 @@
     }
   ],
-  "timestamp": "2025-01-28T15:12:27Z",
+  "timestamp": "2025-01-29T22:47:46Z",
```
(not in scope of this PR): I wonder if we can replace these with placeholders? It's not really useful to have these change every time they're regenerated.

johnstcn commented on provisioner/terraform/testdata/calling-module/calling-module.tfstate.json:
```diff
@@ -26,7 +26,7 @@
             ],
             "env": null,
-            "id": "c559c8be-3b52-444a-bf51-81d270002ec6",
+            "id": "14f0eb08-1bdb-4d48-ab20-e06584ee5b68",
```
(not in scope of this PR): Same here, it's annoying that these get regenerated and cause diffs. Could we replace these with placeholders?

johnstcn commented on provisioner/terraform/resources.go:
```diff
@@ -239,6 +256,44 @@ func ConvertState(ctx context.Context, modules []*tfjson.StateModule, rawGraph s
+				volumesMonitors []*proto.VolumeResourceMonitor
+			)
+
+			for _, resource := range attrs.ResourcesMonitoring {
+				for _, memoryResource := range resource.Memory {
+					if memoryMonitor != nil {
+						// We do not allow multiple memory monitor
+						return nil, xerrors.New("multiple memory monitors are not allowed")
+					}
+					memoryMonitor = &proto.MemoryResourceMonitor{
+						Enabled:   memoryResource.Enabled,
+						Threshold: memoryResource.Threshold,
+					}
+				}
+			}
+
+			monitoredPaths := map[string]struct{}{}
+
+			for _, resource := range attrs.ResourcesMonitoring {
+				for _, volume := range resource.Volumes {
+					if _, exists := monitoredPaths[volume.Path]; exists {
+						return nil, xerrors.Errorf("monitoring the same volume path multiple times is not allowed: %q", volume.Path)
+					}
+					volumesMonitors = append(volumesMonitors, &proto.VolumeResourceMonitor{
+						Path:      volume.Path,
+						Enabled:   volume.Enabled,
+						Threshold: volume.Threshold,
+					})
+				}
+			}
```
I don't see any test cases that cover these code paths? We should add some into `provisionerdserver_test.go`. See the tests for `app_slug`.

johnstcn reviewed (commented):
Looks mostly fine to me, but missing some test cases around the unhappy code paths added to `ConvertState()`.

defelmnq commented on coderd/rbac/roles_test.go:
```diff
@@ -777,6 +777,36 @@ func TestRolePermissions(t *testing.T) {
+			AuthorizeMap: map[bool][]hasAuthSubjects{
+				true: {owner},
+				false: {
+					memberMe, orgMemberMe, otherOrgMember,
```
I am not really sure how to configure RBAC for my needs - I'll let the PR opened until I have enough in the agent one so I can validate RBAC are well configured if good for you.

johnstcn replied:
I think we will need `memberMe` and `orgMemberMe` here, although it will become apparent once we add an API endpoint or similar that tries to read this data.

Adding or extending some existing tests should make it apparent what we need here.

defelmnq committed:
improve database indexes

defelmnq commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -1391,13 +1391,29 @@ func (q *querier) FavoriteWorkspace(ctx context.Context, id uuid.UUID) error {
 }
 
+func (q *querier) FetchMemoryResourceMonitorsByAgentID(ctx context.Context, agentID uuid.UUID) (database.WorkspaceAgentMemoryResourceMonitor, error) {
+	if err := q.authorizeContext(ctx, policy.ActionRead, rbac.ResourceWorkspaceAgentMemoryResourceMonitor); err != nil {
```
I fixed it to only have one RBAC entity 👌

defelmnq committed:
Merge remote-tracking branch 'origin/main' into agent_resource_monitoring

defelmnq committed:
merge main and rename migration

defelmnq committed:
fix resources volumes allocation

defelmnq committed:
regenerate

defelmnq committed:
rename fixture

defelmnq committed:
use agent_id uuid

defelmnq committed:
work on tests

defelmnq committed:
work on tests

defelmnq committed:
add tests on provisionerserver

defelmnq commented:
Fixes : 
- Added foreign key & primary key @ db.
- changed RBAC entities to have one for the resources_monitor instead of one for memory and one for volume.
- Removed validation that will be done in tf-provider instead.
- Improved testing coverage on provisionerserver

dannykopping commented on coderd/database/migrations/000289_agent_resource_monitors.up.sql:
```diff
@@ -0,0 +1,16 @@
+	enabled 	boolean 					NOT NULL,
+	threshold 	integer 					NOT NULL,
+	created_at 	timestamp with time zone 	NOT NULL,
+	PRIMARY KEY (agent_id)
```
Hhmm, I don't think this will work? The primary key needs to be unique.

https://www.postgresql.org/docs/current/ddl-constraints.html#DDL-CONSTRAINTS-PRIMARY-KEYS

> A primary key constraint indicates that a column, or group of columns, can be used as a **unique identifier** for rows in the table.

---

In general, when you're creating database migrations, it's good practice to perform a few basic operations against them to validate that your design works.

I see what you're trying to do, but I think you're try to be overly efficient.

defelmnq replied:
Can you explain to me how we can have duplicates here ? I maybe am missing some logic. 

As we authorize only one resource monitor of type `memory` here - and the resource monitor is linked to an agent id - I feel like it is doing what we want ?

If this one breaks it means that there is multiple memory resource monitors declared in the terraform configuration.

dannykopping replied:
Hhmm, fair point. It's true that each agent should only have one memory resource monitor.

Please leave a comment in the migration explaining this constraint for future readers without context 👍 
_I should not review PRs before bed 🫤_

dannykopping commented on coderd/provisionerdserver/provisionerdserver_test.go:
```diff
@@ -1874,6 +1874,72 @@ func TestInsertWorkspaceResource(t *testing.T) {
+		err := insert(db, job, &sdkproto.Resource{
+			Name: "something",
+			Type: "aws_instance",
+			Agents: []*sdkproto.Agent{{
```
Try be realistic with your testing here; instead of creating two agents (which is uncommon), create one with both resources monitored.

defelmnq replied:
I removed the first agent - just tried to cumulate multiple cases ✅ 
Should be good.

dannykopping replied:
Ah sorry, I see what you were trying to do.
I missed that your second case had a memory component to it.
_Again, I shouldn't review PRs before bed._

defelmnq committed:
remove unrequired test case

johnstcn reviewed (approved):
As the existing tests pass, I'm in favour of merging this and addressing further issues as they arise (unless @dannykopping has objections). The existing PR has become large enough as well (discounting the changes to generated resources) that further changes here may become unwieldy to review. 

Don't forget to check the migration number before merging!

defelmnq committed:
Merge remote-tracking branch 'origin/main' into agent_resource_monitoring
